File: app/src/main/java/com/example/matt/concoctioncrafter/BrewDayFragment.java
/* Copyright Matthew Coughlin 2018, 2019 */

package com.example.matt.concoctioncrafter;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.example.matt.concoctioncrafter.data.Recipe;
import com.example.matt.concoctioncrafter.data.RecipeParcelable;

import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import io.reactivex.disposables.Disposable;

public class BrewDayFragment extends Fragment {
    private Button _startBoilButton;
    private TextView _recipeName;
    private TextView _hop_type1;
    private TextView _hop_type2;
    private TextView _hop_type3;
    private TextView _hop_type4;
    private TextView _amount1;
    private TextView _amount2;
    private TextView _amount3;
    private TextView _amount4;
    private TextView _additionTime1;
    private TextView _additionTime2;
    private TextView _additionTime3;
    private TextView _additionTime4;
    private TextView _alcoholContent;
    private Disposable _recipeSubscription;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getActivity() != null) {
            _recipeSubscription = ((MainActivity) getActivity()).getRecipe().subscribe(this::importRecipe);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        ViewGroup rootView = (ViewGroup) inflater.inflate(R.layout.brew_day_fragment, container, false);

        _startBoilButton = rootView.findViewById(R.id.start_boil_button);
        _startBoilButton.setOnClickListener(v -> Toast.makeText(getContext(), "This feature is coming soon!", Toast.LENGTH_SHORT).show());
        _recipeName = rootView.findViewById(R.id.name);
        _hop_type1 = rootView.findViewById(R.id.hop_type1);
        _hop_type2 = rootView.findViewById(R.id.hop_type2);
        _hop_type3 = rootView.findViewById(R.id.hop_type3);
        _hop_type4 = rootView.findViewById(R.id.hop_type4);
        _amount1 = rootView.findViewById(R.id.amount1);
        _amount2 = rootView.findViewById(R.id.amount2);
        _amount3 = rootView.findViewById(R.id.amount3);
        _amount4 = rootView.findViewById(R.id.amount4);
        _additionTime1 = rootView.findViewById(R.id.time1);
        _additionTime2 = rootView.findViewById(R.id.time2);
        _additionTime3 = rootView.findViewById(R.id.time3);
        _additionTime4 = rootView.findViewById(R.id.time4);
        _alcoholContent = rootView.findViewById(R.id.actual_ac);

        if (getArguments() != null) {
            final RecipeParcelable recipeParcelable = getArguments().getParcelable(MainActivity.RECIPE_KEY);

            if (recipeParcelable != null) {
                final Recipe recipe = recipeParcelable.getRecipe();

                if (recipe != null) {
                    importRecipe(recipe);
                }
            }
        }

        return rootView;
    }

    @Override
    public void onDestroy() {
        _recipeSubscription.dispose();
        super.onDestroy();
    }

    public Button getStartBoilButton() {
        return _startBoilButton;
    }

    public void setStartBoilButton(final CharSequence text) {
        _startBoilButton.setText(text);
    }

    public String getRecipeName() {
        return _recipeName.getText().toString();
    }

    public void setRecipeName(final String name) {
        _recipeName.setText(name);
    }

    public String getHopType1() {
        return _hop_type1.getText().toString();
    }

    public void setHopType1(final String type) {
        _hop_type1.setText(type);
    }

    public String getHopType2() {
        return _hop_type2.getText().toString();
    }

    public void setHopType2(String hop_type2) {
        _hop_type2.setText(hop_type2);
    }

    public String getHopType3() {
        return _hop_type3.getText().toString();
    }

    public void setHopType3(final String hop_type3) {
        _hop_type3.setText(hop_type3);
    }

    public String getHopType4() {
        return _hop_type4.getText().toString();
    }

    public void setHopType4(final String hop_type4) {
        _hop_type4.setText(hop_type4);
    }

    public String getAmount1() {
        return _amount1.getText().toString();
    }

    public void setAmount1(final String amount1) {
        _amount1.setText(amount1);
    }

    public String getAmount2() {
        return _amount2.getText().toString();
    }

    public void setAmount2(final String amount2) {
        _amount2.setText(amount2);
    }

    public String getAmount3() {
        return _amount3.getText().toString();
    }

    public void setAmount3(final String amount3) {
        _amount3.setText(amount3);
    }

    public String getAmount4() {
        return _amount4.getText().toString();
    }

    public void setAmount4(final String amount4) {
        _amount4.setText(amount4);
    }

    public String getAdditionTime1() {
        return _additionTime1.getText().toString();
    }

    public void setAdditionTime1(final String additionTime1) {
        _additionTime1.setText(additionTime1);
    }

    public String getAdditionTime2() {
        return _additionTime2.getText().toString();
    }

    public void setAdditionTime2(final String additionTime2) {
        _additionTime2.setText(additionTime2);
    }

    public String getAdditionTime3() {
        return _additionTime3.getText().toString();
    }

    public void setAdditionTime3(final String additionTime3) {
        _additionTime3 = _additionTime3;
    }

    public String getAdditionTime4() {
        return _additionTime4.getText().toString();
    }

    public void setAdditionTime4(final String additionTime4) {
        _additionTime4.setText(additionTime4);
    }

    private void importRecipe(final Recipe recipe) {
        setRecipeName(recipe.get_recipeName());

        setHopType1(recipe.get_hops().get(0).getName());
        setAmount1(Float.toString(recipe.get_hops().get(0).getAmount()));
        setAdditionTime1(Integer.toString(recipe.get_hops().get(0).getAdditionTime_min()));
        setHopType2(recipe.get_hops().get(1).getName());
        setAmount2(Float.toString(recipe.get_hops().get(1).getAmount()));
        setAdditionTime2(Integer.toString(recipe.get_hops().get(1).getAdditionTime_min()));
        setHopType3(recipe.get_hops().get(2).getName());
        setAmount3(Float.toString(recipe.get_hops().get(2).getAmount()));
        setAdditionTime3(Integer.toString(recipe.get_hops().get(2).getAdditionTime_min()));
        setHopType4(recipe.get_hops().get(3).getName());
        setAmount4(Float.toString(recipe.get_hops().get(3).getAmount()));
        setAdditionTime4(Integer.toString(recipe.get_hops().get(3).getAdditionTime_min()));
    }

    public float getAlcoholContent() {
        return Float.valueOf(_alcoholContent.getText().toString());
    }

    public void setAlcoholContent(final float alcoholContent) {
        _alcoholContent.setText(String.format(Locale.getDefault(), "%.2f", alcoholContent));
    }
}

File: app/src/main/java/com/example/matt/concoctioncrafter/BrewDayFragment.kt
/* Copyright Matthew Coughlin 2018, 2019 */

package com.example.matt.concoctioncrafter

import android.app.AlertDialog
import android.os.Bundle
import android.os.CountDownTimer
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.fragment.app.Fragment
import com.example.matt.concoctioncrafter.MainActivity.Companion.NOTIFICATION_CHANNEL
import com.example.matt.concoctioncrafter.data.Hop
import com.example.matt.concoctioncrafter.data.Recipe
import io.reactivex.disposables.Disposable
import java.util.*
import java.util.concurrent.TimeUnit

class BrewDayFragment : Fragment() {
    private var _startBoilButton: Button? = null
    private var _recipeName: TextView? = null
    private var _hopList: LinearLayout? = null
    private var _alcoholContent: TextView? = null
    private var _timeRemaining: TextView? = null
    private var _recipeSubscription: Disposable? = null
    private var _boilTimer: CountDownTimer? = null
    private var _remainingSeconds: Long = 0
    private var _hops: List<Hop>? = null

    private var recipeName: String
        get() = _recipeName!!.text.toString()
        set(name) {
            if (_recipeName != null)
                _recipeName!!.text = name
            else
                Log.w("Brew_Day_Fragment", "Recipe Name view is null")
        }

    private var alcoholContent: Float
        get() = java.lang.Float.valueOf(_alcoholContent!!.text.toString())
        set(alcoholContent) {
            _alcoholContent!!.text = String.format(Locale.getDefault(), "%.2f", alcoholContent)
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (activity != null) {
            _recipeSubscription = (activity as MainActivity).recipeSubject.subscribe({ recipe ->
                restoreRecipeViews(recipe)
            }, { throwable -> Log.e("Brew_Day_Fragment", "Failed to get the recipe", throwable) })
        }

        if (savedInstanceState != null) {
            _remainingSeconds = savedInstanceState.getLong("REMAINING_TIME")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.brew_day_fragment, container, false) as ViewGroup

        _startBoilButton = rootView.findViewById(R.id.start_boil_button)
        _startBoilButton?.setOnClickListener {
            //Toast.makeText(context, "This feature is coming soon!", Toast.LENGTH_SHORT).show()
            when (_startBoilButton?.text) {
                getString(R.string.start_boil) -> startBoil(TimeUnit.MINUTES.toSeconds(60))
                "Stop Boil" -> stopBoil()
                else -> stopBoil()
            }
        }
        _recipeName = rootView.findViewById(R.id.name)
        _hopList = rootView.findViewById(R.id.hop_info_list)
        _alcoholContent = rootView.findViewById(R.id.actual_ac)
        _timeRemaining = rootView.findViewById(R.id.time_remaining)

        return rootView
    }

    override fun onDestroy() {
        _recipeSubscription?.dispose()
        super.onDestroy()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putLong("REMAINING_TIME", _remainingSeconds)
        super.onSaveInstanceState(outState)
    }

    override fun onStart() {
        if (_remainingSeconds > 0) {
            startBoil(_remainingSeconds)
        }
        super.onStart()
    }

    override fun onStop() {
        if (_boilTimer != null) {
            _boilTimer?.cancel()
        }
        super.onStop()
    }

    /**
     * Restore the recipe information.
     */
    private fun restoreRecipeViews(recipe: Recipe?) {
        if (recipe != null) {
            recipeName = recipe.recipeName
            _hops = recipe.hops
            setHopViews(_hops as ArrayList<Hop>)
        }
    }

    private fun setHopViews(hops: List<Hop>?) {
        _hopList!!.removeAllViews()

        if (hops != null) {
            for (hop in hops) {
                val newRow = layoutInflater.inflate(R.layout.hop_info_row, activity!!.findViewById(R.id.hop_info_list), false)
                newRow.findViewById<TextView>(R.id.hop_name).text = hop.name
                newRow.findViewById<TextView>(R.id.hop_amount).text = "%.2f".format(hop.amount_oz)
                newRow.findViewById<TextView>(R.id.hop_time).text = if (hop.additionTime_min != -1) hop.additionTime_min.toString() else ""

                if (_hopList!!.parent != null) (_hopList!!.parent as ViewGroup).removeView(newRow)
                _hopList!!.addView(newRow)
            }
        }
    }

    private fun startBoil(boilTimeSeconds: Long) {
        _startBoilButton?.text = getString(R.string.stop_boil)
        _timeRemaining?.visibility = View.VISIBLE

        _boilTimer = object : CountDownTimer(TimeUnit.SECONDS.toMillis(boilTimeSeconds), 1000) {
            override fun onFinish() {
                val builder: AlertDialog.Builder? = activity.let { AlertDialog.Builder(it) }
                builder?.setTitle(R.string.boil_finished)
                        ?.setMessage(R.string.click_to_dismiss)
                        ?.setPositiveButton(R.string.ok) { dialog, _ ->
                            stopBoil()
                            dialog.dismiss()
                        }
                        ?.show()
            }

            override fun onTick(millisUntilFinished: Long) {
                _remainingSeconds = millisUntilFinished / 1000

                if (_hops != null) {
                    for (hop in _hops!!) {
                        if (_remainingSeconds == hop.additionTime_min * 60L) {
                            val id: Int = (Math.random() * 1000).toInt()

                            val builder = NotificationCompat.Builder(activity!!.applicationContext, NOTIFICATION_CHANNEL)
                                    .setSmallIcon(R.mipmap.ic_text_launcher_square)
                                    .setContentTitle(getString(R.string.add_title, hop.name))
                                    .setContentText(getString(R.string.add_message, hop.amount_oz))
                                    .setPriority(NotificationCompat.DEFAULT_ALL)

                            with(NotificationManagerCompat.from(activity!!.applicationContext)) {
                                notify(id, builder.build())
                            }

                            activity.let { AlertDialog.Builder(it) }
                                    .setTitle(getString(R.string.add_title, hop.name))
                                    ?.setMessage(getString(R.string.add_message, hop.amount_oz))
                                    ?.setPositiveButton(R.string.ok) { dialog, _ ->
                                        dialog.dismiss()
                                        with(NotificationManagerCompat.from(activity!!.applicationContext)) {
                                            cancel(id)
                                        }
                                    }
                                    ?.show()
                        }
                    }
                }

                _timeRemaining?.text = getString(R.string.remaining_time,
                        TimeUnit.MILLISECONDS.toMinutes(millisUntilFinished),
                        _remainingSeconds - (_remainingSeconds / 60 * 60))
            }
        }.start()
    }

    private fun stopBoil() {
        _startBoilButton?.text = getString(R.string.start_boil)
        _timeRemaining?.visibility = View.GONE
        _boilTimer?.cancel()
        _remainingSeconds = 0
    }
}
